python/cudf/cudf/core/reshape.py

```diff
@@ -1160,3 +1160,271 @@ def _length_check_params(obj, columns, name):
                 f"length of the columns being "
                 f"encoded ({len(columns)})."
             )
+
+
+def _get_names(arrs, names, prefix: str = "row"):
+    """
+    Generates unique names for rows/columns
+    """
+    if names is None:
+        names = []
+        for i, arr in enumerate(arrs):
+            if isinstance(arr, cudf.Series) and arr.name is not None:
+                names.append(arr.name)
+            else:
+                names.append(f"{prefix}_{i}")
+    else:
+        if len(names) != len(arrs):
+            raise AssertionError("arrays and names must have the same length")
+        if not isinstance(names, list):
+            names = list(names)
+
+    return names
+
+
+def crosstab(
+    index,
+    columns,
+    values=None,
+    rownames=None,
+    colnames=None,
+    aggfunc=None,
+    margins=None,
+    margins_name=None,
+    dropna=None,
+    normalize=None,
+):
+    """
+    Compute a simple cross tabulation of two (or more) factors. By default
+    computes a frequency table of the factors unless an array of values and an
+    aggregation function are passed.
+
+    Parameters
+    ----------
+    index : array-like, Series, or list of arrays/Series
+        Values to group by in the rows.
+    columns : array-like, Series, or list of arrays/Series
+        Values to group by in the columns.
+    values : array-like, optional
+        Array of values to aggregate according to the factors.
+        Requires `aggfunc` be specified.
+    rownames : sequence, default None
+        If passed, must match number of row arrays passed.
+    colnames : sequence, default None
+        If passed, must match number of column arrays passed.
+    aggfunc : function, optional
+        If specified, requires `values` be specified as well.
+    margins : Not supported
+    margins_name : Not supported
+    dropna : Not supported
+    normalize : Not supported
+
+    Returns
+    -------
+    DataFrame
+        Cross tabulation of the data.
+
+    Examples
+    --------
+    >>> a = np.array(["foo", "foo", "foo", "foo", "bar", "bar",
```

Comment:
I tried using the same example as the Pandas API. Couldn't use a cupy array as it does not support non-numerical types. Let me know if we still want to use a different example.

Reply:
Correct, cupy won’t do this but a cudf Series will work. Let’s try that. It’s good to use GPU resident data because it helps emphasize the performance value of keeping data on device and avoiding unnecessary copies.
```suggestion
    >>> a = cudf.Series(["foo", "foo", "foo", "foo", "bar", "bar",
```

Reply:
Done. Also had to add a cudf dtype util function which wraps over a pandas util function since the latter does not work with cudf Series.